lib/plugin.js

```diff
@@ -130,72 +128,80 @@ const bind = function(name, pluginBind) {
           return deps.push(el.promise);
         });
       });
-      waitFor = Promise.all(deps);
+      await Promise.all(deps);
     }
-    return waitFor
-      .then(function() {
-        $item.removeClass('emit');
-        let bindPromise = pluginBind($item, item);
-        if (!bindPromise || (typeof(bindPromise.then) === 'function')) {
-          bindPromise = Promise.resolve(bindPromise);
-        }
-        // This is where the "bind results" promise for the current item is stored
-        return $item[0].promise = bindPromise;
-      }).then(function() {
-        // If the plugin has the needed callback, subscribe to server side events
-        // for the current page
-        if (window.plugins[name].processServerEvent) {
-          console.log('listening for server events', $item, item);
-          return forward.init($item, item, window.plugins[name].processServerEvent);
-        }
-      }).catch(e => console.log('plugin emit: unexpected error', e));
-  };
+
+    try {
+
+      $item.removeClass('emit');
+      let bindPromise = $item[0].promise = (async function () {
+        return pluginBind($item, item);
+      })()
+
+      await bindPromise
+
+      // If the plugin has the needed callback, subscribe to server side events
+      // for the current page
+      if (window.plugins[name].processServerEvent) {
+        console.log('listening for server events', $item, item);
+        forward.init($item, item, window.plugins[name].processServerEvent);
+      }
+    } catch (e) {
+      console.log('plugin emit: unexpected error', e)
+    }
+  }
   return fn;
 };
 
-plugin.wrap = function(name, p) {
+plugin.wrap = wrap
+function wrap (name, p) {
   p.emit = emit(p.emit);
   p.bind = bind(name, p.bind);
   return p;
 };
 
-plugin.get = (plugin.getPlugin = function(name, callback) {
+plugin.get = (plugin.getPlugin = async function(name, callback) {
+  if (callback == null) { callback = function() {}; }
   if (window.pluginSuccessor[name]) {
     wiki.log('plugin successor', name, window.pluginSuccessor[name]);
     name = window.pluginSuccessor[name]; 
   }
-  if (loadingScripts[name]) { return loadingScripts[name].then(callback); }
-  loadingScripts[name] = new Promise(function(resolve, _reject) {
-    if (window.plugins[name]) { return resolve(window.plugins[name]); }
-    return getScript(`/plugins/${name}/${name}.js`, function() {
-      let p = window.plugins[name];
+  if (window.plugins[name]) {
```

Comment:
This might be something that was already there in the previous code...

There are a lot of guard clauses checking for `window.plugins[name]` but I can't figure out where that gets assigned. Is that happening in the caller instead of in this file somewhere? (Or am I just missing it?)

Reply:
I did change this a hair, so it doesn't go into the loading scripts mechanism for plugins that are already there. The default plugins all get added to window.plugins in a separate file https://github.com/fedwiki/wiki-client/blob/master/lib/plugins.coffee#L6-L12, and then non-builtin plugins either call `wiki.registerPlugin` from `plugin.registerPlugin` on the last line of this file, or they simply add themselves to `window.plugins` manually.

So the early checks in this file are looking for plugins that have been previously loaded, and the later checks are testing if a plugin has been successfully loaded by seeing if it added itself to `window.plugins`, which also means a plugin can't do anything asynchronous before adding itself to the plugins object or this code thinks that it didn't load successfully. (this part isn't a change, but worth looking at after using import to load plugins lands)